### ISSUE TITLE
Section API

### ISSUE THREAD

**Issue by christopher-besch** [ManimCommunity/manim#2096]
All these changes have already been implemented here: [christopher-besch/manim_web_presenter](https://github.com/christopher-besch/manim_web_presenter)
Everything I'm going to state here has already been documented there (in a better and more exhaustive manner).
You can find a working example showing the benefits of this API [here](https://christopher-besch.github.io/manim_web_presenter/Tutorial).
I used wrapper classes to prevent the creation of a fork.
A direct implementation in Manim would be a lot cleaner though.
This topic has already been discussed on Discord here: [discord.com/channels/581738731934056449/581738732646957057/891746237043310652](https://discord.com/channels/581738731934056449/581738732646957057/891746237043310652)

### The Problem
Currently one has to use the intermediate video files created for each animation in order to stitch together an entire video.
Applying changes in the Python code after the intermediate files have been extracted is quite cumbersome.

### The Solution
The solution is to concatenate multiple animations into slides.
Each slide receives their own output video.
These slides would have to be defined in the code like this:
```py
def construct(self):
    self.next_normal_slide()
    # play some animations
    self.next_normal_slide()
    # play even more animations
```
Every animation between two `self.next_normal_slide()` calls would belong to a single slide.
These calls would optionally take names for the slides.

In addition to the output videos for each slide, an index file (using JSON) would have to be created.
It contains the names for each slides, the name of their respective video and the type of slide.

**Comment by christopher-besch** [comment]
I'm currently working on implementing this API in Manim itself and improving the web interface.
If anyone would like to join my efforts, I'd be very happy.

**Comment by behackl** [comment]
As already mentioned on Discord, I think this is a good idea. It might make sense to flesh out some more implementation-specific details here first, before starting to implement it.

Some thoughts about the user interface:

- There should be just one new method which is able to realize all of the different slide video segments at once. I don't particularly like `.next_TYPE_slide`, I'd rather go in a direction where we have a method `next_slide` that accepts a `type` keyword argument.
- I am also not convinced that the method should have `slide` in its name; when I think of slides I primarily think of something static and not necessarily of a video segment. In the language of video editing, the new method corresponds to cuts; maybe `add_video_cut` or something along these lines might be more intuitive for the context of this particular PR.
- I'm somewhat neutral on the question of CLI flags. I think it would make sense to add a config options for where the video segments and the json manifest should be saved to. I don't have an opinion on whether, in case the new method is called in a scene, the segments are rendered by default (with a CLI flag to disable it if desired), or whether the segments are never rendered by default and a CLI flag needs to be passed if they should be rendered.

What are your thoughts about the interface; what do you feel makes the most sense?

**Comment by christopher-besch** [comment]
1.  I fully agree with using enums for the types.
2.  I agree that `slide` doesn't quite fit for a generic use case. But I don't fancy the idea of defining `cuts`. I think it makes more sense to define `parts` or `sections` that make up the entire video. What matters aren't the cuts but what's between them. Even though logically they provide equivalent information.
3.  I am not a fan of having separate output directories. There will probably be quite some confusion if different outputs end up in different locations.
    I also don't like changing whether or not the JSON file and section-videos should be created based on the existence of one of those `next` calls. In my opinion that's quite a bit too opaque. This can lead to people asking why their JSON file is only sometimes being created. If people don't use any `next` calls, there will only be one output anyways. So the overhead would be minimal and could be optimized into nothingness quite easily. I know that this is an extremely opinionated statement; I simply don't like software pretending to be smarter than the user.
    I don't know how popular this API is going to be; if it ends up being used by everybody, the only useful CLI flag would be one turning the API off for quick debugging. But in the beginning people would probably complain why their rendering times are longer—especially without caching. So it might not be a bad idea to start off with one and remove it later if it gets too annoying.

In addition to that I'd like to add a point raised by @kolibril13. He showed that working with multiple `Scene` classes isn't possible with the current draft API. I think the best way of handling this is by using reflection to find all the defined classes in order of definition and creating one big JSON file for all classes. This could be enabled by using the `--write_all` CLI flag; if you're rendering a single scene, only parts of the JSON would have to be updated.

**Comment by behackl** [comment]
> But I don't fancy the idea of defining cuts. I think it makes more sense to define parts or sections that make up the entire video. What matters aren't the cuts but what's between them.

Sure, cut was just a suggestion. I like your suggestions and am fine with part / section / segment etc.; I just thought that slide was a bit too specialized.

> I am not a fan of having separate output directories. There will probably be quite some confusion if different outputs end up in different locations.

I am not sure I understand; my intention was to group the segments together somehow so that they are not easily confused with "normal" video outputs. Currently, the default location to save output files is `media/videos/module_name/resolution`, and my suggestion would be to save the segments to, for example, `media/videos/module_name/resolution/segments` or `media/segments/module_name/resolution` (maybe with a slight preference for the first one). In any case, independent of what the default value is, it should be possible to change it via a config option.

> I also don't like changing whether or not the JSON file [...]

This is a good argument against automatic detection and for requiring the config option. 👍 

> multiple `Scene` classes

Personally, I'd rather write one manifest file per scene class; it's more in the *spirit* of `--write_all`. This seems like a simpler approach and might be better suited for a first implementation. It does mean you have to read from several manifests to process all segments, but this shouldn't be a problem. What do you think?

**Comment by christopher-besch** [comment]
Yeah, now I agree with you on the output locations. I totally misunderstood your first statement, sorry my bad.

There are multiple reasons in favour of allowing multiple `Scene` classes:
1. Quicker Debugging: If you're working on projects with rendering times in excess of 10min. it might be useful to split that projects into smaller parts. This might not be an issue with better caching.
2. I've never used Jupyter Notebook with Manim; I don't know how it would work with this API. Being able to split a project in multiple scenes might help here.
3. Currently we have multiple types of `Scene` classes. Without allowing multiple classes, it wouldn't be possible to use 2D and 3D in one project. This obviously wouldn't be a problem if this mess of scene classes were to be refactored.
4. I don't like having an entire project in a single `Scene` class. With this, variables would be all over the place. Having different scopes would force some order. I've never worked with bigger Manim projects so I don't know how big of a problem this is. I also don't know if different methods would be enough.

I intend the presentation front end to do quite a bit more than just presenting. It'd be something like a Manim video editor where you can record the video and have a system that synchronises the audio with the animations. This system could take care of handling multiple `Scene` classes. That way the API wouldn't have to support multiple `Scene` classes itself.
What do you think?

Thanks for your opinion!

**Comment by kolibril13** [comment]
> There are multiple reasons in favour of allowing multiple Scene classes.

To 1: Totally agree. It happens very easily that one gets rendering times over 20 minutes for big projects. Also, often I recycle code from previous slides, and use the same variable names. It can happen easily, that variables are then getting overwritten in a wrong way. Therefore: Splitting scenes is a good idea!
To 2: I've worked with it a lot, and it is amazing and convenient because one can see any time the video and the corresponding code close together. I think this presentation feature would go very well with it.
To 4:  Splitting things into smaller parts that are easy to handle is great! I only have the feeling, that in *some* situations, it might be useful to re-use mobjects, that were defined before (so it would be nice to have e.g. 3 slides in one class ). 

Here is what I could imagine would be nice to work with:
The presentation contains sections, and in these sections, one can have snippets.
Stating a new snippet could be done by writing  "next_snippet" and the slide type could be added by the "action" keyword.
Either as a string, or even as a variable, how it was done in the [ArrowTip](https://docs.manim.community/en/stable/reference/manim.mobject.geometry.ArrowTip.html?highlight=arrowtip#arrowtipsshowcase), which would enable auto-completion.
Code could possibly look like this:
```py
class Section1(Presentation):
    def construct(self):
        self.next_snippet(action = "normal_slide", name= "normal slide")
        mob = ...
        self.add(mob,...)
        self.wait()

        self.next_snippet(action = "normal_slide", name= "another slide")
        mob = ...
        self.add(mob,...)
        self.wait()
        
class Section2(Presentation):
    def construct(self):
        self.next_snippet(action = "loop", name= "looped slide")
        mob = ...
        self.add(mob,...)
        self.wait()

        self.next_snippet(action = "normal_side", name= "and here it goes")
        mob = ...
        self.add(mob,...)
        self.wait()
  ```

**Comment by christopher-besch** [comment]
Thanks for your input!
Here are my thoughts on that:

> The presentation contains sections, and in these sections, one can have snippets.

I don't quite understand the difference between your section and snippet.
What do we need yet another subdivision for?
Being allowed to use multiple classes would obviously not replace the `next_whatever` calls. It would only offer some additional structuring.

> Either as a string, or even as a variable, how it was done in the ArrowTip, which would enable auto-completion.

This means using enums which I really like.

> Code could possibly look like this[...]

The real question is how to connect the different classes together.
These are the two options discussed:
1. Use reflection and create one big JSON file when using `--write_all` or
2.  have multiple JSON files, one for each scene. Then the presentation / video recording front end would have to stitch the scenes together.
    I prefer the second option.
    I'm generally not a big fan of reflection and don't like the added complexity; I don't particularly fancy self-aware code :)

    But the first option might make sense if this API were to receive widespread use.

**Comment by kolibril13** [comment]
> I don't quite understand the difference between your section and snippet.
here is a draft:
![image](https://user-images.githubusercontent.com/44469195/135503797-c1e6053a-5694-49e5-818b-8f05934cc520.png)
One section is generated from one class, and in each section, you can have multiple snippets.

> The real question is how to connect the different classes together.

I would also prefer approach 2, as it allows iterating over presentations more quickly. And it would be even possible to quickly add new slides after a few weeks.

What about this approach:
```python
@section_position(num=1)
class Section1(Presentation):
    def construct(self):
   ...

@section_position(num=2)
class Section2(Presentation):
    def construct(self):
    ...
```
Then the decorator can take care that the snippets are written into the correct jsonfile.
The numbering could be done after a pattern 1, 2, 3, ...
When one wants to change the order, one just can change the number.
In jupyter notebooks, I guess one could even do something like `@section_position(num=notebook.get_current_cell_number())`
That would mean: When one wants to change the order of slides, one just changed the order of the cells, which is a very similar behavior one already knows from powerpoint:
![image](https://user-images.githubusercontent.com/44469195/135505720-6c5f1528-2583-4fed-8677-94fd98b2b3a9.png)

**Comment by christopher-besch** [comment]
I still don't see why we need another level of subdividing; where should that information be shown. In the presenter?
I kinda feel like that'd only add more complexity, confusion and points of failure for very little payoff.
We might need a poll to decide this one.

> What about this approach[...]

I don't like having to define the arrangement with numbers like this.
This looks like Word's vs. LaTex's way of doing table of contents.
If you want to for example insert a section, you'd have to alter all the indices from there on.
I don't think these numbers are something the users should be worried about.

There might be a pretty solution for the notebooks but I want something as important as this to work for all users—including terminal addicts like myself.
In any case I don't quite see why we can't just use the order of definition.
This should also work for Jupyter and who in the world would want the classes to be read in a different order anyways?

This would again require some complex reflection.
So the easiest solution—at least in the beginning—would be to define the order in the frontend.
That would mean that the frontend displays all the scenes it can find in the `media` output folder and asks which ones to load and in what order.

Is this acceptable?
